SignalMessaging/profiles/OWSUserProfile.h

```diff
@@ -0,0 +1,90 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import <SignalServiceKit/TSYapDatabaseObject.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef void (^OWSUserProfileCompletion)(void);
+
+@class OWSAES256Key;
+
+extern NSString *const kNSNotificationName_LocalProfileDidChange;
+extern NSString *const kNSNotificationName_OtherUsersProfileWillChange;
+extern NSString *const kNSNotificationName_OtherUsersProfileDidChange;
+
+extern NSString *const kNSNotificationKey_ProfileRecipientId;
+extern NSString *const kNSNotificationKey_ProfileGroupId;
+
+extern NSString *const kLocalProfileUniqueId;
+
+// This class should be completely thread-safe.
+@interface OWSUserProfile : TSYapDatabaseObject
+
+@property (atomic, readonly) NSString *recipientId;
+@property (atomic, readonly, nullable) OWSAES256Key *profileKey;
+@property (atomic, readonly, nullable) NSString *profileName;
+@property (atomic, readonly, nullable) NSString *avatarUrlPath;
+// This filename is relative to OWSProfileManager.profileAvatarsDirPath.
+@property (atomic, readonly, nullable) NSString *avatarFileName;
+
+// This should reflect when either:
+//
+// * The last successful update finished.
+// * The current in-flight update began.
+@property (atomic, readonly, nullable) NSDate *lastUpdateDate;
+
+- (instancetype)init NS_UNAVAILABLE;
+
++ (OWSUserProfile *)getOrBuildUserProfileForRecipientId:(NSString *)recipientId
+                                           dbConnection:(YapDatabaseConnection *)dbConnection;
+
+#pragma mark - Update With... Methods
+
+- (void)updateWithProfileName:(nullable NSString *)profileName
+                   profileKey:(OWSAES256Key *)profileKey
+                avatarUrlPath:(nullable NSString *)avatarUrlPath
+               avatarFileName:(nullable NSString *)avatarFileName
+                 dbConnection:(YapDatabaseConnection *)dbConnection
+                   completion:(nullable OWSUserProfileCompletion)completion;
```

Comment:
The problem with the "update with..." pattern is that if you want to modify just a subset of the properties, you need to generate a separate method for each permutation of possibly modified properties.

Reply:
The only thing I could think of would be to just expose a method that takes the change block, but we'd be putting more responsibility on the caller to do the right thing.

I mean, I guess all the painful work of writing it has been done now anyway. =)

Reply:
> The only thing I could think of would be to just expose a method that takes the change block, but we'd be putting more responsibility on the caller to do the right thing.

That wouldn't even work since the setters are (by design) private to prevent non-updateWith... access